text/0000-native-link-modifiers.md

```diff
@@ -0,0 +1,232 @@
+- Feature Name: `native_link_modifiers`
+- Start Date: 2020-06-12
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Provide an extensible mechanism for tweaking linking behavior of native libraries
+both in `#[link]` attributes (`#[link(modifiers = "+foo,-bar")]`)
+and on command line (`-l static:+foo,-bar=mylib`).
+
+# Motivation
+[motivation]: #motivation
+
+Occasionally some tweaks to linking behavior of native libraries are necessary,
+and currently there's no way to apply them.
+
+For example, some static libraries may need to be linked as a whole archive
+without throwing any object files away because some symbols in them appear to be unused
+while actually having some effect. \
+This RFC introduces modifier `whole-archive` to address this.
+
+In other cases we need to link to a library located at some specific path
+or not matching the default naming conventions. \
+This RFC introduces modifier `verbatim` to pass such libraries to the linker.
+
+This RFC also reformulates the `static-nobundle` linking kind as a modifier `bundle`
+thus providing an opportunity to change the static linking default to non-bundling
+on some future edition boundary, and hopefully unblocking its stabilization.
+
+The generic syntax provides a way to add more such modifiers in the future
+without introducing new linking kinds.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This is an advanced feature not expected to be used commonly,
+see the reference-level explanation if you know that you need some of these modifiers.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Existing syntax of linking attributes and options
+
+- Attributes: `#[link(name = "string", kind = "string", cfg(predicate))]`
+(some components are optional.)
+- Command line options: `-l kind=name:rename` (some components are optional).
+
+## Proposed extensions to the syntax
+
+- Attributes: `#[link(/* same */, modifiers = "+foo,-bar")]`.
+- Command line options: `-l kind:+foo,-bar=name:rename`.
+
+The modifiers are boolean and applied only to the single library specified with `name`. \
+`+` means enable, `-` means disable, multiple options can be separated by commas,
+the last boolean value specified for the given modifier wins. \
```

Comment:
Any reason to do this instead of erroring if multiple are specified?

Reply:
Would it be possible to have a deny-by-default lint in code, then? Command line switches make perfect sense, but if someone includes multiple in the code it's almost certainly a mistake.

Reply:
Even if the multiple specifier case doesn't make sense, it's still very hard to write it accidentally, so making a new lint for it doesn't pull its weight. (There are plenty opportunities to write some unnatural code that doesn't make any sense but is not linted.)
I guess it's easier to make it a hard error then.